pep-0544.txt

```diff
@@ -439,98 +448,123 @@ but a static type checker will handle them specially. Subclassing a protocol
-the existence of ``SupportsClose`` from the example in `definition`_ section::
+Alternatively, one can implement ``SizedAndClosable`` protocol by merging
+the ``SupportsClose`` protocol from the example in `definition`_ section
+with ``typing.Sized``::
 
   from typing import Sized
 
-  class SupportsClose(...): ...  # Like above
+  class SupportsClose(Protocol):
+      def close(self) -> None:
+          ...
 
-  class SizedAndCloseable(Sized, SupportsClose, Protocol):
+  class SizedAndClosable(Sized, SupportsClose, Protocol):
       pass
 
 The two definitions of ``SizedAndClosable`` are equivalent.
 Subclass relationships between protocols are not meaningful when
 considering subtyping, since structural compatibility is
 the criterion, not the MRO.
 
-Note that rules around explicit subclassing are different from regular ABCs,
-where abstractness is simply defined by having at least one abstract method
-being unimplemented. Protocol classes must be marked *explicitly*.
+If ``Protocol`` is included in the base class list, all the other base classes
+must be protocols. A protocol can't extend a regular class, see `rejected`_
+ideas for rationale. Note that rules around explicit subclassing are different
+from regular ABCs, where abstractness is simply defined by having at least one
+abstract method being unimplemented. Protocol classes must be marked
+*explicitly*.
 
 
-Generic and recursive protocols
--------------------------------
+Generic protocols
+-----------------
 
 Generic protocols are important. For example, ``SupportsAbs``, ``Iterable``
 and ``Iterator`` are generic protocols. They are defined similar to normal
 non-protocol generic types::
 
-  T = TypeVar('T', covariant=True)
-
   class Iterable(Protocol[T]):
       @abstractmethod
       def __iter__(self) -> Iterator[T]:
           ...
 
-Note that ``Protocol[T, S, ...]`` is allowed as a shorthand for
-``Protocol, Generic[T, S, ...]``.
+``Protocol[T, S, ...]`` is allowed as a shorthand for
+``Protocol, Generic[T, S, ...]``. Declaring variance is not necessary for
+protocol classes, since it can be inferred from a protocol definition.
+Example::
+
+  class Box(Protocol[T]):
+      content: T
+
+  x: Box[float]
+  y: Box[int]
+  x = y  # This is OK due to inferred covariance of the protocol 'Box'.
+
+
+Recursive protocols
+-------------------
 
 Recursive protocols are also supported. Forward references to the protocol
 class names can be given as strings as specified by PEP 484. Recursive
-protocols will be useful for representing self-referential data structures
+protocols are useful for representing self-referential data structures
 like trees in an abstract fashion::
 
   class Traversable(Protocol):
       leaves: Iterable['Traversable']
 
+Note that for recursive protocols, a class is considered a subtype of
+the protocol in situations where such decision depends on itself.
+Continuing the previous example::
+
+  class SimpleTree:
+      leaves: List['SimpleTree']
```

Comment:
Since `leaves` is a mutable attribute, `SimpleTree` doesn't look compatible with `Traversable`. If the attribute in `Traversable` would be a read-only property, this would be fine.